### DISCUSSION TITLE
Use SpaceMouse at Runtime?

### DISCUSSION THREAD

**Discussion by seltzdesign** [microdee/OpenUnrealSpaceMouse#97] [General]
Just getting started in Unreal after extensively using vvvv (beta and gamma). This might be a total noob question, but:

Can we use the SpaceMouse as input method at runtime, ie. in the compiled game? 

Do we have to use Blueprints for that? Is there a template? It should be in addition to normal mouse controls.

Thanks so much for this plugin by the way, it makes so much nicer to use the Editor using familiar controls. I just find the defaults are all quite opposite to what we are used to. We use Orbiting No Roll, but with 2 of the rotation axis and 1 translation axis inverted. Then you have the default "Object Mode" like it is in Rhino, which is pretty much perfect by default and just super intuitive. They have their own driver built in, so I don't even have the crappy 3DxWare installed. Curious why the defaults here are so different. I saved my settings now to always use again.

**Comment by microdee** [comment]
Hey hey, VVVV ftw!

> Can we use the SpaceMouse as input method at runtime, ie. in the compiled game?

With this plugin yes

> Do we have to use Blueprints for that? Is there a template? It should be in addition to normal mouse controls.

Yes you need to program your own navigation method in BP, but you can do so with Unreal's input mapping abstractions (you have an action and you map multiple device input axes to it). There's 6DoF game template in Unreal iirc, you can just edit the axis mappings in there to include space mice, OR I use this project for testing and development: https://github.com/microdee/UE4-SpaceMouse-TestProject . That includes a pawn which you can pilot 6DoF and has physics collisions during gameplay.

Now I haven't tried my plugin yet with the new Enhanced Input method, which you should use for new projects, but as a device input it should be automatically supported, I think at least.

> Curious why the defaults here are so different

I myself like to express the local-space-camera-delta-transformation with my puck, not the movement of the object / world. Reminiscent a bit like flight stick controls of an airplane. But I understand that every person has a different preference of 3D navigations, so that's why I included the 3x3 matrix mappings. Unfortunately though if you want to add the same options for the users of your application, you need to program that yourself, the runtime part of this plugin just gives you the axis readings and pipes them into Unreal's own input mapping system.

**Comment by seltzdesign** [comment]
Awesome, thanks for that! I will give it a go.

**Reply by tietoafinlandoy** [reply]
@microdee what would it mean/ take to establish the same functionality on MacOS builds? Runtime support of 3D SpaceMouse for packaged game using PC + Mac setup to build the game.

**Reply by microdee** [reply]
You sent me an email about this to which I replied, and a longer response here https://github.com/microdee/UE4-SpaceMouse/issues/4#issuecomment-1866934851

**Comment by tietoafinlandoy** [comment]
Ah, sure. I’ll keep the discussion there…

From: David Mórász ***@***.***>
Date: Thursday, 21. December 2023 at 23.58
To: microdee/UE4-SpaceMouse ***@***.***>
Cc: Tomas Westerholm ***@***.***>, Comment ***@***.***>
Subject: Re: [microdee/UE4-SpaceMouse] Use SpaceMouse at Runtime? (Discussion #97)
Et saa usein sähköpostia lähettäjältä ***@***.*** Lisätietoja siitä, miksi tämä on tärkeää<https://aka.ms/LearnAboutSenderIdentification>

You sent me an email about this to which I replied, and a longer response here #4 (comment)<https://github.com/microdee/UE4-SpaceMouse/issues/4#issuecomment-1866934851>
